File: knn_classifier.py
# +
import numpy as np
import matplotlib.pyplot as plt

from sklearn.datasets import make_blobs


# -

class KNNClassifier:

	def __init__(self, k=5):
		self.k = k

	def fit(self, X, y):
		self.X_train = X
		self.y_train = y

	def euclidean_distance(self, x1, x2):
		return np.sqrt(np.sum(x1 - x2) ** 2)

	def predict(self, X):
		y_pred = [self._predict_single(x) for x in X]
		return np.array(y_pred)

	def _predict_single(self, x):
		# Calculate distances between x and all examples in the training set
		distances = [self.euclidean_distance(x, x_train) for x_train in self.X_train]

		# Get the indices of the k-nearest neighbors
		k_indices = np.argsort(distances)[:self.k]

		# Get the labels of the k-nearest neighbors
		k_nearest_labels = [self.y_train[i] for i in k_indices]

		# Return the most common class label among the k-nearest neighbors
		most_common = np.bincount(k_nearest_labels).argmax()

		return most_common


# Generate sample data
n_samples = 400
n_components = 3

X, y_true = make_blobs(
    n_samples=n_samples, centers=n_components, cluster_std=0.60, random_state=0
)
X = X[:, ::-1]

# Plot init seeds along side sample data
plt.figure(1)
colors = ["#4EACC5", "#FF9C34", "#4E9A06", "m"]


for k, col in enumerate(colors):
    cluster_data = y_true == k
    plt.scatter(X[cluster_data, 0], X[cluster_data, 1], c=col, marker=".", s=10)

model = KNNClassifier(5)
model.fit(X, y_true)

y_pred = model.predict(X)

for k, col in enumerate(colors):
    cluster_data = y_pred == k
    plt.scatter(X[cluster_data, 0], X[cluster_data, 1], c=col, marker=".", s=10)
plt.xticks([])
plt.yticks([])

File: knn_classifier.jl
import CluGen: clugen
using Plots
import StatsBase: countmap


struct KNNClassifier

    k :: Int
    X_train :: Matrix{Float64}
    y_train :: Vector{Int}

    KNNClassifier(k, X, y) = new( k, X, y)

end

euclidean_distance(x1, x2) = sqrt(sum(x1 .- x2).^2)

predict(knn, X) = [predict_single(knn, x) for x in X]

function predict_single(knn, x)

	# Calculate distances between x and all examples in the training set
    distances = [euclidean_distance(x, x_train) for x_train in eachrow(knn.X_train)]

	# Get the indices of the k-nearest neighbors
	k_indices = partialsortperm(distances, 1:knn.k)

	# Get the labels of the k-nearest neighbors
	k_nearest_labels = view(knn.y_train, k_indices)

	# Return the most common class label among the k-nearest neighbors
    most_common = argmax(countmap(k_nearest_labels))

	return most_common

end


# Generate sample data
dimension = 2
n_samples = 400
n_components = 3

options = ([1, 0], pi / 8, [20, 10], 10, 1, 1.5)

o = clugen(dimension, n_components, n_samples, options...)

X = o.points
y_true = o.clusters

scatter(X[:,1], X[:,2], group = y_true)

model = KNNClassifier(5, X, y_true)

y_pred = predict(model, X)

scatter(X[:,1], X[:,2], group = y_true)
